### DISCUSSION TITLE
Difficult to convert a v6 `MantineThemeOverride` to v7 `MantineTheme`

### DISCUSSION THREAD

**Discussion by japem** [mantinedev/mantine#5499] [Feedback]
Because of the breaking change to how theming works from Mantine 6 to 7, there is no good way to take a `MantineThemeOverride` in v6 and use it as a `MantineTheme` in v7:

- The `theme` prop in `MantineProvider` in v6 takes in a type `MantineThemeOverride`, and in v7 it takes in a `MantineTheme`
- There is no way in v6 to merge a theme override with the default theme and turn it into a `MantineTheme`
- Writing a utility to convert this is not trivial due to the (semantic?) differences in the types between the two versions. Additionally, the `DeepPartial` type used throughout the `MantineThemeOverride` type causes differences in the types of properties between the `MantineThemeOverride` and `MantineTheme` even within the same version.
- Thus there is no way to easily & reliably migrate from the theme used by a site using Mantine 6 and make it work for Mantine 7.

It would be great if there was an official utility function to convert from a v6 `MantineThemeOverride` to a v7 `MantineTheme` (and maybe some other things as well, for example going back and exporting the theme-merge function in a new minor version of Mantine 6).

I understand that not everything in a Mantine 6 theme has a corollary in Mantine 7. So it would also be helpful to provide a detailed migration guide for each of these properties (for example, color scheme, which has been lifted out of the `MantineTheme` type entirely).

In general it seems like Mantine 7 introduces an unnecessarily onerous migration path and I'm curious why there were so many breaking changes to fundamental parts of the library. It seems like lots of these changes were made without any consideration to how frustrating it might be for a consumer to accommodate it.

My company uses Mantine for several sites and migrating each of them to use Mantine 7 looks like it's going to be far more work than it really should need to be. The least this library could do is provide some utilities to make the migration easier.

**Comment by rtivital** [comment]
There weren't any changes to MantineProvider theme type, it still uses MantineThemeOverride type – https://github.com/mantinedev/mantine/blob/master/packages/%40mantine/core/src/core/MantineProvider/MantineProvider.tsx#L18

I do not understand what the issue is.

**Reply by japem** [reply]
Interesting, my IDE types it as `PartialObjectDeep<MantineTheme, {}> | undefined` (maybe this is equivalent to `MantineThemeOverride`?) but I do see under the hood that the type is actually a `MantineThemeOverride`. So, that's solved at least, thank you.

Still though, the migration guide for a v6 theme (override) to a v7 theme (override) could be better documented. For example:

- the color scheme, this is at least mentioned offhand in the migration guide.
- style functions for components (`MantineThemeComponent.styles`) have different signatures (and are typed as `any` in v7, which causes runtime errors when carrying over a function from v6)
- the new `VariantColorsResolver` - is this net-new or is there a corollary in v6 in a different place?
- the `MantineThemeColors` type is subtly different
- `transitionTimingFunction` has gone away
- ...etc
